### PR TITLE
Allow Opt out of all automated validation messages and emails

### PR DESCRIPTION
**Pull request by fiennyangeln**
#635

### CODE REVIEW

**Comment by bgirardot** [comment]
@fiennyangeln Testing now....

**Comment by bgirardot** [comment]
@fiennyangeln Taking longer than I thought as I do not have a good SMTP env set up for testing, but I am working on it :)

**Comment by bgirardot** [comment]
@fiennyangeln Hi, before I spend any more time testing this, can you please confirm it works as described in my last message on the issue?

Specifically that it disables any automated tasking manager internal message from validation ?

This is not just about not getting an email notification of a message or a validation message, this is about not get any automated messages from the validation process.

The other kinds of messages where some one uses the @username should still work as they do now and should still be sent via email if they have entered an email address.

Basically we want to turn off any validation message, regardless of if that validation message would generate an email or not.

Please let me know if this is how it works, from just a quick glance it looks like you missed my last message on the issue thread and I feel pretty bad, that is my fault for not @fiennyangeln you.

**Comment by bgirardot** [comment]
@fiennyangeln Yes, I see, the whole issue says "Email" but this issue has changed since it was originally created 8 months ago.

More folks have complained about getting the validation messages in general, not just the emails.

Even the internal Tasking Manager messages they get they are complaining about.

That is why I think it might be best, easiest to just opt out of any auto generated validation messages and then that would automatically stop the the emails because there would be no message sent.

I will update the issue to reflect this and not just say email anymore.

**Comment by fiennyangeln** [comment]
Ok Blake. For now, this is what I turn off : 
-Sends mapper a thank you, after their task has been marked as valid 
-send a canned message to anyone @'d in a comment
-Send alert to user if they were @'d in a chat message

**Comment by bgirardot** [comment]
@fiennyangeln Yes, I did just test this and it looks like the the TM internal messages are disabled for basically everything if you ... unselect your new option.

So it seems like it would work exactly correctly if that only applied to the automatically generated validation message.

no automatically generated validation tasking manager internal message
no email notification about the automatically generated validation message

But yes, we always want the internal tasking manger message for @username mentions in any chat and we always want to send the email notifications for those messages.

If they do not wish to receive any emails they can just remove their email address and then they still get the internal tasking manager messages, but will not get the emails of course, that is how it works now. But @username mentions always need to generate the internal message and if they have provided their email addres, generate the email notice to them.

**Comment by fiennyangeln** [comment]
@bgirardot ok I modified it. pls review

**Comment by bgirardot** [comment]
@fiennyangeln I am still testing this, just thought of another use case that needs testing, but my testing so far of your revised pull request shows it is working as it should, which is great :)

Can I ask you to please change the code itself to be more descriptive?

your variable names now are misleading because they all talk about 'email' but this is not really about email, it is about opting out of automated validation messages, not getting the emails about them is just a side effect of not getting any automated validation messages at all.

Does that make sense?

In looking over the code, I think you could literally do a find/replace to replace email_notification with validation_messages

**Comment by fiennyangeln** [comment]
@bgirardot changed.

**Comment by NibesKing** [comment]
Hope it works.

**Comment by bgirardot** [comment]
@smit1678 @ethan-nelson As I said above, I feel like this is well tested via the front end, it just needs a code review and to be deployed carefully as it does a database change via the migration.

**Comment by smit1678** [comment]
> deployed carefully as it does a database change via the migration

👍 We can run checks on the staging database once it is merged and deployed

**Comment by smit1678** [comment]
Just noticed the conflicts. @fiennyangeln can you review those conflicts and resolve? I think they may be related to the changes you made in https://github.com/hotosm/tasking-manager/commit/c204c7faad0d86d935173a58f4347d123c87c259

**Comment by fiennyangeln** [comment]
@smit1678 ok will do!

**Comment by fiennyangeln** [comment]
@smit1678 done!!:D

**Comment by bgirardot** [comment]
@smit1678 @dakotabenjamin I recall this being ready to merge into dev. The only hold up was that it did a migration of the db so we might expect a hiccup. Probably not, but could happen. I did the db updgrades and downgrades for testing and it worked fine, as did the db migrate up and down, that worked fine as well. Can one of you please merge it to develop at your convenience.?

**Comment by dakotabenjamin** [comment]
Do you want this in the the master PR as well?

**Comment by bgirardot** [comment]
@dakotabenjamin I would say this could be in the master pr as well. whatever works best. I am fine with leaving it out too.

**Comment by bgirardot** [comment]
@dakotabenjamin I feel that way about all of them. a few are ready to go I think, but if they do not make it, they do not make it, no big deal.

**Comment by JesseWeinstein** [comment]
The master PR this was in got reverted, but I can't figure out why. Can someone link me to the discussion?
